# App Store Review

Using data from both the Apple Store and the Google Play Store, the goal of this project is to help our make believe company determine which app attributes lead to more downloads and ratings. This fictional company makes free apps in English only and just makes money from ad revenue.  

Both datasets can be found on Kaggle:  
[Google Play Store Data](https://www.kaggle.com/lava18/google-play-store-apps/home)  
[Apple Store Data](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home)

In [1]:
from csv import reader
opened_file = open('AppleStore.csv', encoding='utf-8')
read_file = reader(opened_file)
apple = list(read_file)

opened_file = open('googleplaystore.csv', encoding='utf-8')
read_file = reader(opened_file)
google = list(read_file)

In [2]:
googlecols = google[0]
applecols = apple[0]

apple = apple[1:]
google = google[1:]

In [3]:
def explore_data(dataset, start, end, rows_and_columns = False):
    #For easier to view data (using list format, not pandas)
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row, end='\n\n')
        
    if rows_and_columns:
        print('Rows:' + str(len(dataset)))
        print('Columns:' + str(len(dataset[0])))

In [4]:
print(googlecols, end = '\n\n')
explore_data(google, 0, 3, rows_and_columns = True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']

Rows:10841
Columns:13


In [5]:
print(applecols, end = '\n\n')
explore_data(apple, 0, 3, rows_and_columns = True)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']

['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']

['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']

Rows:7197
Columns:17


### Available variables

Some google play store variables that might be helpful:  
- Category
- Rating
- Reviews
- Size
- Installs
- Type
- Content Rating
- Genres
- Last Updated

In other words, most of them. The names for the Apple Play Store variables are less clear, but we can still get some idea:  
- track_name  
- size_bytes   
- rating_count_tot  
- rating_count_ver (current version only)  
- user_rating  
- user_rating_ver (current version only)
- cont_rating  
- prime_genre  


### Data Accuracy

Before checking accuracy, we'll first want to remove non-English apps and those that are not free. There would probably be some value in reviewing the same things for apps by price to explore differences, but let's focus on the main goal.

We can also remove duplicates and check for errors.

In [6]:
#Let's start with Google - since we're working with lists and not data frames, let's check the lengths for consistency
googlelengths = {}
for row in google:
    rowlength_str = str(len(row))
    if rowlength_str in googlelengths:
        googlelengths[rowlength_str] += 1
    else:
        googlelengths[rowlength_str] = 1
print(googlelengths)

{'13': 10840, '12': 1}


Aha! Looks like we have an inconsistent row. Where is it?

In [7]:
rowid = 0
for row in google:
    rowid += 1
    if len(row) == 12:
        fixrow = rowid - 1
    
print(fixrow)
        

10472


In [8]:
print(googlecols, end = '\n\n')
explore_data(google, fixrow, fixrow +2)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']



OK, two things to fix. For one, the App is missing Category. Second, it's missing the Genre. Checking the Google Play Store, it seems this is a Lifestyle app. Knowing this, we could fix it, but this is a Dataquest project and they suggest removing the row, so let's do what they do and delete it.


In [9]:
del(google[fixrow])

In [10]:
#Confirm the row is now gone:
google[fixrow][0]

'osmino Wi-Fi: free WiFi'

Now, the same check for the Apple Store.

In [11]:
applelengths = {}
for row in apple:
    rowlength_str = str(len(row))
    if rowlength_str in applelengths:
        applelengths[rowlength_str] += 1
    else:
        applelengths[rowlength_str] = 1
print(applelengths)

{'17': 7197}


Looks good. Next, let's check for duplicates.

### Checking for duplicates

In [12]:
#Again, google first

def listdupcheck(a_list, dup_idx = 0):
    #Function to check for duplicates in a particular field of a list of lists
    #a_list is the list
    #dup_idx is the index of the field in the inner lists to check for duplicates
    apps = []
    dupapps  = []
    for row in a_list:
        if row[dup_idx] in dupapps:
            next
        elif row[dup_idx] in apps:
            dupapps.append(row[dup_idx])
        else:
            apps.append(row[dup_idx])
    return(apps, dupapps)
    
dupcheck_results = listdupcheck(google)
apps = dupcheck_results[0]
dupapps = dupcheck_results[1]

print(len(apps))   
print(len(dupapps))


9659
798


798 apps appear more than once in the data! Let's separate the data, then keep whichever instance of the duplicate app has the most reviews (and so is probably the most recent). 9,659 is the number of unique apps. When we clean up our duplicates, the length of the cleaned up data should be 9,659.

In [13]:
googleunique = []
googledup = []
for row in google:
    if row[0] in dupapps:
        googledup.append(row)
    else:
        googleunique.append(row)

print(len(googleunique))
print(len(googledup))

8861
1979


In [14]:
#The number of reviews is at index 3
maxratings = {}
for row in googledup:
    if row[0] not in maxratings:
        maxratings[row[0]] = row[3]
    else:
        maxratings[row[0]] = max(row[3], maxratings[row[0]])
from pprint import pprint

sample = 0
for key in maxratings:
    pprint(key)
    pprint(maxratings[key])
    sample += 1
    if sample > 4:
        break
    

'Coloring book moana'
'974'
'Mcqueen Coloring pages'
'65'
'UNICORN - Color By Number & Pixel Art Coloring'
'8264'
'Textgram - write on photos'
'295237'
'Wattpad 📖 Free Books'
'2915189'


In [15]:
googledup_recent = []
already_added = []
for row in googledup:
    if row[3] == maxratings[row[0]] and row[0] not in already_added:
        googledup_recent.append(row)
        already_added.append(row[0])

print(len(googledup_recent))

798


798 is the number we were looking for. Now we'll combine the separated datasets (always unique and unique-from-duplicates) and make sure we get a list of length 9,659.

In [16]:
google = googleunique + googledup_recent

print(len(google))

9659


Cool. Now again, but for Apple.

In [17]:
appledup_results = listdupcheck(apple, 2)
apps = appledup_results[0]
dupapps = appledup_results[1]
print(len(apps))
print(len(dupapps))

7195
2


Ah, just 2. We can use the same strategy and keep the one with more ratings.

In [18]:
#rating_count_tot is index 6

maxratings = {}
for row in apple:
    if row[2] in dupapps:
        ratings = row[6]
        if row[2] in maxratings:
            maxratings[row[2]] = max(maxratings[row[2]], ratings)
        else:
            maxratings[row[2]] = ratings

#Only a couple
pprint(maxratings)

{'Mannequin Challenge': '668', 'VR Roller Coaster': '67'}


In [19]:
apple_fixed = []
for row in apple:
    if row[2] not in maxratings:
        apple_fixed.append(row)
    elif row[2] in maxratings and row[6] == maxratings[row[2]]:
        apple_fixed.append(row)

print(len(apple_fixed))

7195


In [20]:
apple = apple_fixed.copy()

OK, now duplicate values have been removed from the google and apple data sets, and we've fixed an issue with a missing value. Next, we'll focus on English. Only one of the data sets have a language field we might use, and we'd have to interpret it (still might do that). But for now, we can use `ord()` to identify any titles with characters that are not standard English (anything from 0 to 127 returned from `ord()`). This simple filter would also pick up emoji's in names, or things like copyright or trademark symbols. To get a bit closer to the truth, we'll require at least 4 characters with `ord()` greater than 127.

### Filtering for language

In [21]:
def isenglish(str):
    countnonenglish = 0
    for character in str:
        if ord(character) > 127:
            countnonenglish += 1
    if countnonenglish > 3:
        return(False)
    else:
        return(True)

print(isenglish('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(isenglish('doggy dog'))
print(isenglish('Instachat 😜'))

False
True
True


In [22]:
#Google
google_eng = []
for row in google:
    if isenglish(row[0]):
        google_eng.append(row)
        
#Apple
apple_eng = []
for row in apple:
    if isenglish(row[2]):
        apple_eng.append(row)

print(len(google_eng))
print(len(apple_eng))

9614
6181


OK. Fewer rows, but not a dramatic cut. Apple lost a lot more than google when checking for foreign, which is mildly interesting. Finally, to cut down to free apps:

### Free Apps Only

In [23]:
#Google (Type index 6 - Free, Paid, or Nan - just one Nan)
google = [] #Lets recover the original name
for row in google_eng:
    if row[6] == 'Free':
        google.append(row)

#Apple (Price index 5)
apple = [] #Same. Recovering name.
for row in apple_eng:
    if float(row[5]) > 0:
        apple.append(row)
        
print(len(google))
print(len(apple))


8861
2961


Wow, Apple sure had a higher proportion of paid apps than google. The Apple dataset was reduced by more than half. Google, less than 10%. Data cleaning complete.

### Identifying winners

Our plan is to identify Apps that are popular with both Apple and Android users. The steps we'll take:  

1. Add minimal app to Google Play Store
2. If it gets a good response, develop it further
3. After 6 months, if app is profitable, build iOS version and add to Apple Store.

Let's start by identifying common which genres tend to have the most apps.

### Genre frequency tables

First, some functions ind index indentifiers.

In [24]:
ggcat_idx = 1  #google category
gggenre_idx = 9  #google genre
agenre_idx = 12  #apple genre

def freq_table(dataset, index, percents = False):
    freq_dict = {}
    for row in dataset:
        if row[index] in freq_dict:
            freq_dict[row[index]] += 1
        else:
            freq_dict[row[index]] = 1
    
    if percents:
        for key in freq_dict:
            freq_dict[key] /= len(dataset)
            freq_dict[key] *= 100
            freq_dict[key] = round(freq_dict[key], 2)
            
    return(freq_dict)

def sort_dict(a_dict):
    listtosort = []
    for key in a_dict:
        listtosort.append((a_dict[key], key))
    
    return(sorted(listtosort, reverse = True))


Let's just look at the top 10 for each group

In [25]:
#Prepare sorted lists
googlecat = sort_dict(freq_table(google, ggcat_idx, percents = True))
googlegenre = sort_dict(freq_table(google, gggenre_idx, percents = True))
applegenre = sort_dict(freq_table(apple, agenre_idx, percents = True))

#Google Categories
print('Google Categories')
for entry in googlecat[0:10]:
    print(entry[1] + " : " + str(entry[0]))
 
#Google Genres
print('\nGoogle Genres')
for entry in googlegenre[0:10]:
    print(entry[1] + " : " + str(entry[0]))
        
#Apple Genres
print('\nApple Genres')
for entry in applegenre[0:10]:
    print(entry[1] + " : " + str(entry[0]))

Google Categories
FAMILY : 18.93
GAME : 9.69
TOOLS : 8.45
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.52
SPORTS : 3.4
PERSONALIZATION : 3.32

Google Genres
Tools : 8.44
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.52
Sports : 3.46
Personalization : 3.32

Apple Genres
Games : 51.27
Education : 9.86
Entertainment : 6.59
Photo & Video : 6.11
Utilities : 4.46
Productivity : 3.78
Health & Fitness : 3.38
Music : 2.4
Lifestyle : 1.62
Weather : 1.38


#### Apple

A full half of the free, English language apps on the Apple Store are games. That's huge. Entertainment also factors highly. Fun is definitely key in the Apple top 10. The remaining top 10 are virtually all more "useful" apps. Education, naturally, but also productivity/utilities/health & fitness, with music being the only other app we might put in the "just for fun" category. However, that 51% Games is hard to ignore (without looking at download figures, anyway... getting there).

#### Google

If we look at categories, Family and Games came out on top, though even combined they're still under 30% of the total - there's more spread than in the Apple Store. Having said that, the rest of the top 10 looks a lot like Apple, with "useful" apps rounding out the remainder outside of entertainment (with the exception of sports). Hard to say where Lifestyle belongs, but it does appear in all 3 lists.

Now, we'll look at the average number of ratings in each genre.

In [35]:
ratingtot_idx = 6
apple_rbg = {} #ratingsbygenre

for row in apple:
    ratingcount = int(row[ratingtot_idx])
    genre = row[12]
    if genre in apple_rbg:
        newratingcount = apple_rbg[genre][0] + ratingcount
        genre_n         = apple_rbg[genre][1] + 1
        apple_rbg[genre] = (newratingcount, genre_n)
    else:
        apple_rbg[genre] = (ratingcount, 1)

apple_avgrating = {}
for key in apple_rbg:
    ratingsum = apple_rbg[key][0]
    ratingn   = apple_rbg[key][1]
    apple_avgrating[key] = round(float(ratingsum) / float(ratingn),0)
    
sort_dict(apple_avgrating) #Average NUMBER of ratings, to be clear.

[(6696.0, 'Games'),
 (4043.0, 'Business'),
 (3872.0, 'News'),
 (3248.0, 'Weather'),
 (2759.0, 'Music'),
 (2722.0, 'Shopping'),
 (2680.0, 'Health & Fitness'),
 (2532.0, 'Photo & Video'),
 (2400.0, 'Reference'),
 (2248.0, 'Productivity'),
 (2132.0, 'Entertainment'),
 (1327.0, 'Utilities'),
 (1309.0, 'Catalogs'),
 (1175.0, 'Navigation'),
 (903.0, 'Lifestyle'),
 (883.0, 'Finance'),
 (664.0, 'Medical'),
 (641.0, 'Education'),
 (603.0, 'Travel'),
 (580.0, 'Food & Drink'),
 (393.0, 'Social Networking'),
 (320.0, 'Book'),
 (254.0, 'Sports')]

Well, games seem like a winner on Apple's App Store (based solely on how many there are and how many ratings each of them tends to get. Entertainment and Photo & Video also seem to do well. Let's check out google ratings, too.

### Google number of ratings

Rather than reviews for Google, we'll use installs... not for any particular reason other than that we don't have a nice figure to work with so we get a minor bit of string practice. Oh, and vague estimation.

In [69]:
googlecat = freq_table(google,ggcat_idx)

googleavgratings = {}
for category in googlecat:
    total = 0.0
    len_category = 0.0
    for row in google:
        category_app = row[ggcat_idx]
        if category_app == category:
            installs = float(row[5].replace('+','').replace(',',''))
            total += installs
            len_category += 1.0
    avg_installs = int(installs / len_category)
    googleavgratings[category] = avg_installs

sort_dict(googleavgratings)
#print(googlecat)
    

[(628930, 'VIDEO_PLAYERS'),
 (175438, 'ART_AND_DESIGN'),
 (120481, 'LIBRARIES_AND_DEMO'),
 (117647, 'ENTERTAINMENT'),
 (116414, 'GAME'),
 (90909, 'FOOD_AND_DRINK'),
 (70422, 'WEATHER'),
 (60606, 'DATING'),
 (52631, 'BOOKS_AND_REFERENCE'),
 (48076, 'EDUCATION'),
 (40322, 'MAPS_AND_NAVIGATION'),
 (36630, 'HEALTH_AND_FITNESS'),
 (34013, 'PERSONALIZATION'),
 (28901, 'LIFESTYLE'),
 (24154, 'TRAVEL_AND_LOCAL'),
 (21186, 'SOCIAL'),
 (18181, 'COMICS'),
 (14492, 'PRODUCTIVITY'),
 (13698, 'HOUSE_AND_HOME'),
 (13351, 'TOOLS'),
 (5963, 'FAMILY'),
 (5025, 'SHOPPING'),
 (4032, 'NEWS_AND_MAGAZINES'),
 (3831, 'PHOTOGRAPHY'),
 (3484, 'COMMUNICATION'),
 (3048, 'FINANCE'),
 (2457, 'BUSINESS'),
 (1886, 'BEAUTY'),
 (332, 'SPORTS'),
 (160, 'MEDICAL'),
 (17, 'PARENTING'),
 (7, 'EVENTS'),
 (1, 'AUTO_AND_VEHICLES')]

Well, in the Google Play store, things don't line up very well. But, it's not totally fair. Because google buckets the installs, any category with a lot of apps may also have a lot of apps in the lowest possible category, which drags the average down a lot. Any category with fewer apps, especially when a couple of apps might dominate the category, will have its contributions weighted a bit higher.

Other than Games, which comes in 5th for average installs, many of the top categories in # of apps factor fairly low in terms of average installs. The next closest are personalization and lifestyle.

